core/src/cache.rs

```diff
@@ -220,6 +221,65 @@ impl<E> CacheError<E> {
     }
 }
 
+/// Input data usually comes from files on the file system, but there are also
+/// lots of cases where we want to synthesize other kinds of inputs.
+///
+/// Note that a `SourcePath` does not uniquely identify a cached input:
+/// - Some functions (like [`Cache::add_file`]) add a new cached input unconditionally.
+/// - [`Cache::get_or_add_file`] will add a new cached input at the same `SourcePath` if
+///   the file on disk was updated.
+///
+/// The equality checking of `SourcePath` only affects [`Cache::replace_string`], which
+/// overwrites any previous cached input with the same `SourcePath`.
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub enum SourcePath {
+    /// A file at the given path.
+    ///
+    /// Note that this does not need to be a real file on the filesystem: it could still
+    /// be loaded from memory by, e.g, [`Cache::add_string`].
```

Comment:
I experimented with that. Maybe it makes sense to do that privately, but I think this enum is the correct one for the public API, because this way `add_string` can add a `SourceKind::Memory` of any `SourcePath` variant, and `add_file` can only add a `SourcePath::Path` with `SourceKind::Filesystem`. These are all the valid combinations of `SourcePath` and `SourceKind`. (I thought a little about how to encode the types internally to only allow these valid combinations, but it didn't seem worth the trouble.)

If we separated out `SourcePath::Path` into a `SourcePath::RealPath` and `SourcePath::MemoryPath`, someone could call `add_string` with `SourcePath::RealPath` which wouldn't make sense.

Reply:
Ah, because the in-memory source for the LSP still maps to a path on the disk? I admit I don't remember very well which is the use of the in-memory sources.

Reply:
I'm not sure I understood what "maps to" means, but an in-memory `SourcePath::Path` variant can be resolved as the target of an import statement. In fact, it's the only variant with this property -- I'll add that to the comment.

The main point of an in-memory source is that we don't try to auto-reload it from disk. Basically an in-memory `SourcePath::Path` acts as an in-memory override of the file on disk at that path.

Reply:
Ok, then that indeed makes sense that they relate to a path. I wrongly thought that path didn't make sense for an in-memory import.